### PR TITLE
Us1/5 signup system and login system

### PR DESCRIPTION
**Pull request by SukhyAnand**
Implemented functionality of both login and signup system with node.js server integration, MongoDB database connection and server side token based user authentication.

### CODE REVIEW

**Comment by abhinawsarang** [comment]
1. This is not how you write commit messages `
US5-Tasks-24,25,26,27,28,29,30,32,34,32,33,36,37,38,39,40,54,55,56` it not readable.
2. I don't see any point of refactoring everything (Client side code) just to add server. Why can't we have a directory name "server"  inside NumberNinja root?

**Comment by abhinawsarang** [comment]
Closing this PR and reopening it to trigger the CI test.

**Comment by SukhyAnand** [comment]
> Can you please see why these empty files are here `favicon.ico` `.gitkeep`

Both of them are auto-generated files and it is advisable to keep them even if they are empty. They serve an important purpose in the code. Read the below lines to get more clarity about the files:

1) "Many web browsers have a facility where, if someone bookmarks your website (ie puts it in their "Favorites" menu), it associates a particular icon with your website in the bookmark. It tries to obtain that icon by first requesting for "favicon.ico" from the directory of your web page. If it cannot find such a file, it will try to obtain it from the root directory of your website, failing which, it will simply use a default icon for the bookmark."

2) "Git cannot add a completely empty directory. People who want to track empty directories in Git have created the convention of putting files called .gitkeep in these directories. The file could be called anything; Git assigns no special significance to this name". Since we may make empty directories in our project during the development process for creating component namespaces, keeping .gitkeep is useful.

File: NumberNinjaServer/routes/index.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+var express = require('express');
+var router = express.Router();
+
+/* GET home page. */
+router.get('/', function(req, res, next) {
+  res.render('index', { title: 'Express' });
```
Comment (abhinawsarang) [inline NumberNinjaServer/routes/index.js]: You want the title to be "Express" ?

File: NumberNinjaServer/routes/users.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
+var express = require('express');
+var router = express.Router();
+var User = require('../models/user');
+var jwt = require('jsonwebtoken');
+
+router.post('/register',  function(req,res,next){
```
Comment (abhinawsarang) [inline NumberNinjaServer/routes/users.js]: Comments on major functions.

File: src/app/app-routing.module.ts [hunk 3]
```diff
@@ -1,11 +0,0 @@
-import { NgModule } from '@angular/core';
-import { Routes, RouterModule } from '@angular/router';
-
-
-const routes: Routes = [];
-
-@NgModule({
-  imports: [RouterModule.forRoot(routes)],
-  exports: [RouterModule]
-})
```
Comment (abhinawsarang) [inline src/app/app-routing.module.ts]: Don't we need this file ? Why it is deleted ?

Comment (SukhyAnand) [reply src/app/app-routing.module.ts]: it's moved to another file called routes.ts which serves the same purpose.

File: NumberNinja/src/app/auth/auth.guard.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { Injectable } from '@angular/core';
+import { CanActivate, ActivatedRouteSnapshot, RouterStateSnapshot, Router } from '@angular/router';
+import { UserService } from '../shared/user.service';
+//import { Observable } from 'rxjs/Observable';
+
+@Injectable()
+export class AuthGuard implements CanActivate {
+  constructor(private userService : UserService, private router : Router){}
+  
+  canActivate(
```
[the remark this answers is not in the document]
Comment (SukhyAnand) [reply NumberNinja/src/app/auth/auth.guard.ts]: Done.

File: NumberNinja/src/app/auth/auth.interceptor.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import 'rxjs/add/operator/do';
+import { Injectable } from "@angular/core";
+import { Router } from "@angular/router";
+
+@Injectable()
+export class AuthInterceptor implements HttpInterceptor {
+
+    constructor(private router: Router) { }
+
+    intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
```
[the remark this answers is not in the document]
Comment (SukhyAnand) [reply NumberNinja/src/app/auth/auth.interceptor.ts]: Done.

File: NumberNinja/src/app/auth/auth.interceptor.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        if (req.headers.get('No-Auth') == "True")
+            return next.handle(req.clone());
+
+        if (localStorage.getItem('userToken') != null) {
+            console.log("Hello");
```
[the remark this answers is not in the document]
Comment (SukhyAnand) [reply NumberNinja/src/app/auth/auth.interceptor.ts]: Done.

File: NumberNinja/src/app/auth/auth.interceptor.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (SukhyAnand) [reply NumberNinja/src/app/auth/auth.interceptor.ts]: Testing this won't be possible at this stage. The interceptor is basically used to pass your authentication token along with the header of the HTTP requests which are being sent to the server.
At this point of time I have only 2 apis for signup and signin. Once this change is integrated with the dashboards or student, teacher and admin, the spec file will be developed accordingly to make calls to other server apis.

File: NumberNinja/src/app/shared/user.service.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import { Observable } from 'rxjs';
+import 'rxjs/add/operator/map';
+import { User } from './user.model';
+
+@Injectable()
+export class UserService {
+  readonly rootUrl = 'http://localhost:3000';
+  constructor(private http: HttpClient) { }
+
+  registerUser(user: User) {
```
[the remark this answers is not in the document]
Comment (SukhyAnand) [reply NumberNinja/src/app/shared/user.service.ts]: Done.

File: NumberNinja/src/app/user/signup/signup.component.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import { UserService } from '../../shared/user.service';
+
+@Component({
+  selector: 'app-sign-up',
+  templateUrl: './signup.component.html',
+  styleUrls: ['./signup.component.css']
+})
+export class SignUpComponent implements OnInit {
+  user: User;
+  emailPattern = "^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}$";
```
[the remark this answers is not in the document]
Comment (SukhyAnand) [reply NumberNinja/src/app/user/signup/signup.component.ts]: Done.

File: NumberNinjaServer/app.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+var cookieParser = require('cookie-parser');
+var logger = require('morgan');
+
+var indexRouter = require('./routes/index');
+var usersRouter = require('./routes/users');
+
+var app = express();
+
+// Add mongoose 
+var mongoose = require('mongoose');
```
[the remark this answers is not in the document]
Comment (SukhyAnand) [reply NumberNinjaServer/app.js]: Done.

File: NumberNinjaServer/routes/index.js (hunk 1 above)
Comment (SukhyAnand) [reply NumberNinjaServer/routes/index.js]: This is just the default api to check whether express server is accessible. Hence let the title be Express.

File: NumberNinja/src/app/routes.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
+import { Routes } from '@angular/router'
+import { HomeComponent } from './home/home.component';
+import { UserComponent } from './user/user.component';
+import { SignUpComponent } from './user/signup/signup.component';
+import { SignInComponent } from './user/signin/signin.component';
+import { AuthGuard } from './auth/auth.guard';
+
+export const appRoutes: Routes = [
+    { path: 'home', component: HomeComponent,canActivate:[AuthGuard] },
```
[the remark this answers is not in the document]
Comment (SukhyAnand) [reply NumberNinja/src/app/routes.ts]: Done.

File: NumberNinjaServer/routes/users.js (hunk 2 above)
Comment (SukhyAnand) [reply NumberNinjaServer/routes/users.js]: Done.

File: NumberNinjaServer/routes/users.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+/* POST API to register user details to users collection of 
+ * MongoDB database and also check whether the user's email
+ * is registered before.
+ */
+router.post('/register',  function(req,res,next){
+  var user = new User({
+    email: req.body.Email,
+    username: req.body.UserName,
+    password: User.hashPassword(req.body.Password),
+    firstname: req.body.FirstName,
```
Comment (Sagar-k93) [inline NumberNinjaServer/routes/users.js]: Use camel case for variables.

File: NumberNinjaServer/routes/users.js [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    lastname: req.body.LastName,
+    role: null,
+    creation_dt: Date.now()
```
Comment (Sagar-k93) [inline NumberNinjaServer/routes/users.js]: Please follow camelCase format for variable use.

File: NumberNinjaServer/routes/users.js (hunk 11 above)
Comment (abhinawsarang) [reply NumberNinjaServer/routes/users.js]: Done.

File: NumberNinjaServer/routes/users.js (hunk 12 above)
Comment (abhinawsarang) [reply NumberNinjaServer/routes/users.js]: Done.

**Pull request merged**